web/src/queries/network.ts

```diff
@@ -0,0 +1,330 @@
+ */
+const accessPointsQuery = () => ({
+  queryKey: ["network", "accessPoints"],
+  queryFn: async (): Promise<AccessPoint[]> => {
+    const accessPoints = await fetchAccessPoints();
+    return accessPoints.map(AccessPoint.fromApi).sort((a, b) => (a.strength < b.strength ? -1 : 1));
+  },
+  // FIXME: Infinity vs 1second
+  staleTime: 1000,
+});
+
+/**
+ * Hook that builds a mutation to add a new network connection
+ *
+ * It does not require to call `useMutation`.
+ */
+const useAddConnectionMutation = () => {
+  const queryClient = useQueryClient();
+  const query = {
+    mutationFn: (newConnection: Connection) =>
+      addConnection(newConnection.toApi())
+        .then(() => applyChanges())
+        .catch((e) => console.log(e)),
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ["network", "connections"] });
+      queryClient.invalidateQueries({ queryKey: ["network", "devices"] });
+      queryClient.invalidateQueries({ queryKey: ["network", "accessPoints"] });
+    },
+  };
+  return useMutation(query);
+};
+/**
+ * Hook that builds a mutation to update a network connection
+ *
+ * It does not require to call `useMutation`.
+ */
+const useConnectionMutation = () => {
+  const queryClient = useQueryClient();
+  const query = {
+    mutationFn: (newConnection: Connection) =>
+      updateConnection(newConnection.toApi())
+        .then(() => applyChanges())
+        .catch((e) => console.log(e)),
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ["network", "connections"] });
+      queryClient.invalidateQueries({ queryKey: ["network", "devices"] });
+    },
+  };
+  return useMutation(query);
+};
+
+/**
+ * Hook that builds a mutation to remove a network connection
+ *
+ * It does not require to call `useMutation`.
+ */
+const useRemoveConnectionMutation = () => {
+  const queryClient = useQueryClient();
+  const query = {
+    mutationFn: (name: string) =>
+      deleteConnection(name)
+        .then(() => applyChanges())
+        .catch((e) => console.log(e)),
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: ["network", "connections"] });
+      queryClient.invalidateQueries({ queryKey: ["network", "devices"] });
+    },
+  };
+  return useMutation(query);
+};
+
+/**
+ * Returns selected Wi-Fi network
+ */
+const selectedWiFiNetworkQuery = () => ({
+  // queryKey: ["network", "wifi", "selected"],
+  // TODO: use right key, once we stop invalidating everything under network
+  queryKey: ["wifi", "selected"],
+  queryFn: async () => {
+    return Promise.resolve({ ssid: null, needsAuth: null });
+  },
+  staleTime: Infinity,
+});
+
+const useSelectedWifi = (): { ssid?: string; needsAuth?: boolean } => {
+  const { data } = useQuery(selectedWiFiNetworkQuery());
+  return data || {};
+};
+
+const useSelectedWifiChange = () => {
+  type SelectedWifi = {
+    ssid?: string;
+    hidden?: boolean;
+    needsAuth?: boolean;
```

Comment:
so selected wifi can be empty? it looks kind of strange. I expect that at least ssid is mandatory.

Reply:
It's not that strange. You can connect to a _hidden_ Wi-Fi network.

Reply:
Yeah, I know how hidden network works :) But the problem is that this is a "UI query" for keeping which network is/was selected by the user in order to show the proper information about it: the connection form or its configuration/state and actions.

But the user could select a "hidden network", for which the UI does not know the SSID in advance. That's why the ssid here is optional. Maybe a matter of naming or a matter or typing improvement (I'm just landing in the typed languages world :P)

How would you improve it?

Reply:
at least document that info in the type as it is confusing for reader. So mention that for hidden network ssid is not known in advance. I am not sure if you can have something in ts like that hidden is true or ssid is specified.

Reply:
>  I am not sure if you can have something in ts like that hidden is true or ssid is specified.

Interesting point. I'm going to see if that it's possible. Thanks a ton!

Reply:
Argh! Actually is a bit more tricky than I though because I forgot to mention a small detail in previous explanations: we're updating the information in the query in a moment in which we don't know the ssid but only that the authentication failed. So, we assumed that authentication failed for the selected connection. See https://github.com/openSUSE/agama/pull/1519/files#diff-f1b1879261cce23d74bb1716c65fbbcb487ee658acf074133fc6d6def7288736R223-R263

So, to make it work without TypeScript complaints the changes would be something like 


```diff
diff --git a/web/src/queries/network.ts b/web/src/queries/network.ts
index 00754d6d..d8fecc7a 100644
--- a/web/src/queries/network.ts
+++ b/web/src/queries/network.ts
@@ -193,16 +193,25 @@ const useSelectedWifi = (): { ssid?: string; needsAuth?: boolean; hidden?: boole
 };
 
 const useSelectedWifiChange = () => {
-  type SelectedWifi = {
-    ssid?: string;
-    hidden?: boolean;
-    needsAuth?: boolean;
+  type VisibleWifi = {
+    ssid: string;
+    hidden?: false | never;
   };
 
+  type HiddenWifi = {
+    ssid?: never;
+    hidden: true;
+  };
+
+  type KnownWifi = VisibleWifi & { needsAuth?: boolean };
+
+  type SelectedWifi = KnownWifi | HiddenWifi;
+
   const queryClient = useQueryClient();
 
   const mutation = useMutation({
-    mutationFn: async (data: SelectedWifi): Promise<SelectedWifi> => Promise.resolve(data),
+    mutationFn: async (data: Partial<SelectedWifi>): Promise<Partial<SelectedWifi>> =>
+      Promise.resolve(data),
     onSuccess: (data: SelectedWifi) => {
       queryClient.setQueryData(["wifi", "selected"], (prev: SelectedWifi) => ({
         ssid: prev.ssid,
```

On one hand, from the _typing POV_ would be more clear that a hidden wifi is _unknown_ until the user enters the ssid and became it in a _KnownWifi_. On the other hand... although safer and robust code looks a bit complicated. What do you think?